lib/rules/length-zero-no-unit/index.js

```diff
@@ -2,201 +2,162 @@
-					return [' ', ',', ')', '/'].includes(char);
-				});
-
-				// If no next break was found, this value ends at the end of the string
-				const valueWithZeroEnd =
-					nextValueBreakIndex === -1 ? value.length : nextValueBreakIndex + valueWithZeroStart;
-
-				const valueWithZero = value.slice(valueWithZeroStart, valueWithZeroEnd);
-				const parsedValue = valueParser.unit(valueWithZero);
-
-				if (!parsedValue || (parsedValue && !parsedValue.unit)) {
-					return;
-				}
-
-				if (parsedValue.unit.toLowerCase() === 'fr') {
-					return;
-				}
-
-				// Add the indexes to ignorableIndexes so the same value will not
-				// be checked multiple times.
-				_.range(valueWithZeroStart, valueWithZeroEnd).forEach((i) => (ignorableIndexes[i] = true));
-
-				// Only pay attention if the value parses to 0
-				// and units with lengths
-				if (
-					Number.parseFloat(valueWithZero) !== 0 ||
-					!keywordSets.lengthUnits.has(parsedValue.unit.toLowerCase())
-				) {
-					return;
-				}
-
-				if (context.fix) {
-					fixPositions.unshift({
-						startIndex: valueWithZeroStart,
-						length: valueWithZeroEnd - valueWithZeroStart,
-					});
-
-					return;
-				}
-
-				report({
-					message: messages.rejected,
-					node,
-					index: valueWithZeroEnd - parsedValue.unit.length,
-					result,
-					ruleName,
-				});
+			const parsedValue = valueParser(value);
+
+			parsedValue.walk((valueNode) => {
+				return check(node, getIndex(node), valueNode);
 			});
 
-			if (fixPositions.length) {
-				fixPositions.forEach((fixPosition) => {
-					if (node.type === 'atrule') {
-						// Use `-1` for `@` character before each at rule
-						const realIndex =
-							fixPosition.startIndex - node.name.length - node.raws.afterName.length - 1;
+			if (needsFix) {
+				set(node, parsedValue.toString());
+			}
+		}
+
+		function checkDecl(node, value, set, getIndex) {
+			needsFix = false;
 
-						node.params = replaceZero(node.params, realIndex, fixPosition.length);
-					} else {
-						const realIndex = fixPosition.startIndex - node.prop.length - node.raws.between.length;
+			const { prop } = node;
 
-						node.value = replaceZero(node.value, realIndex, fixPosition.length);
-					}
-				});
+			if (isLineHeight(prop)) return;
+
+			if (optionsMatches(secondary, 'ignore', 'custom-properties') && isCustomProperty(prop))
+				return;
+
+			const parsedValue = valueParser(value);
+
+			parsedValue.walk((valueNode, valueNodeIndex, valueNodes) => {
+				if (isLineHeightValue(node, valueNodes, valueNodeIndex)) return;
+
+				return check(node, getIndex(node), valueNode);
+			});
+
+			if (needsFix) {
+				set(node, parsedValue.toString());
 			}
 		}
+
+		root.walkAtRules((atRule) =>
+			checkAtRule(atRule, getAtRuleParams(atRule), setAtRuleParams, atRuleParamIndex),
+		);
+		root.walkDecls((decl) =>
+			checkDecl(decl, getDeclarationValue(decl), setDeclarationValue, declarationValueIndex),
+		);
```

Comment:
[suggestion] IMO, it seems not intuitive to me to pass arguments except for a node to `checkAtRule()` and `checkDecl()`, since `getAtRuleParams()` or `getDeclarationValue()` are used only once.

Here is my suggestion to use the utility function in `checkAtRule()` and `checkDecl()`:

```suggestion
		root.walkAtRules(checkAtRule);
		root.walkDecls(checkDecl);
```

or,

```suggestion
		root.walkAtRules((atRule) => checkAtRule(atRule));
		root.walkDecls((decl) => checkDecl(decl));
```

For example, `checkDecl()` will be:

```js
function checkDecl(node) {
	needsFix = false;

	// ...

	const value = getDeclarationValue(node);
	const index = declarationValueIndex(node);
	const parsedValue = valueParser(value);

	parsedValue.walk((valueNode, valueNodeIndex, valueNodes) => {
		if (isLineHeightValue(node, valueNodes, valueNodeIndex)) return;

		return check(node, index, valueNode);
	});

	// ...
}
```

If doing so, we can lazily evaluate `getDeclarationValue(node)` or `declarationValueIndex(node)`. The same applies to `checkAtRule()`.

Reply:
Excellent suggestion, thanks.